### PR TITLE
[SYSTEMML-1428] Fixed maxpooling functions for padding > 0

### PR DESCRIPTION
**Pull request by niketanpansare**
@bertholdreinwald @dusenberrymw

### CODE REVIEW

**Comment by dusenberrymw** [comment]
@niketanpansare Unfortunately, I'm still experiencing test failures with this PR using the tests added in #434.  Can you grab that version of the `nn` library and run the tests to observe the issues?

**Comment by niketanpansare** [comment]
@dusenberrymw interesting, I compared this PR with the results returned by CuDNN. Here is the code that I used for testing: https://github.com/niketanpansare/systemml-cudnn-test ... In particular, can you please check if the tests satisfy the assumption "do not treat pad as zero" (see https://github.com/apache/incubator-systemml/pull/437/files#diff-ec9eaa0bc47052d8d2acbba33fbd8f89R570)  which is consistent with CuDNN ?

**Comment by dusenberrymw** [comment]
Indeed, I was not aware of that assumption, but I see that is true within other implementations as well, albeit the docs aren't always correct :).  I'll make a few updates to my code & tests and then see where we are at.

File: src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixDNN.java [hunk 1]
```diff
@@ -433,16 +433,14 @@ private static void fillIndexesArray(ConvolutionParameters params) {
 		params.start_indexes_w = new int[params.Q];
 		params.end_indexes_w = new int[params.Q];
 		for (int p = 0; p < params.P; p++) {
-			int start_index_h = p * params.stride_h - params.pad_h;
-			final int end_index_h = Math.min(start_index_h + params.R, params.H);
-			start_index_h = Math.max(start_index_h, 0);
+			int start_index_h = p * params.stride_h - params.pad_h; // Math.max(p * params.stride_h - params.pad_h, 0);
+			int end_index_h = start_index_h + params.R; // Math.min(start_index_h + params.R, params.H);
```
Comment (dusenberrymw) [inline src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixDNN.java]: Can you remove the commented out code before merging?

File: src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixDNN.java [hunk 2]
```diff
@@ -433,16 +433,14 @@ private static void fillIndexesArray(ConvolutionParameters params) {
...
 			params.start_indexes_h[p] = start_index_h;
 			params.end_indexes_h[p] = end_index_h;
 		}
 		for (int q = 0; q < params.Q; q++) {
-			int start_index_w = Math.max(q * params.stride_w - params.pad_w, 0);
-			int end_index_w = Math.min(start_index_w + params.S, params.W);
-			start_index_w = Math.max(start_index_w, 0);
+			int start_index_w =  q * params.stride_w - params.pad_w; // Math.max(q * params.stride_w - params.pad_w, 0);
+			int end_index_w = start_index_w + params.S; // Math.min(start_index_w + params.S, params.W);
```
Comment (dusenberrymw) [inline src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixDNN.java]: Can you remove the commented out code before merging?

File: src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixDNN.java [hunk 3]
```diff
@@ -899,7 +909,7 @@ public static void maxpooling(MatrixBlock input, MatrixBlock outputBlock, Convol
 		//post-processing: maintain nnz
 		outputBlock.recomputeNonZeros();
 	}
-
+
```
Comment (dusenberrymw) [inline src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixDNN.java]: nitpick: can you remove this whitespace?

File: src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixDNN.java [hunk 4]
```diff
@@ -919,8 +929,11 @@ private static void doPooling(int n, ConvolutionParameters params) throws DMLRun
 				final int inOffset1 = inOffset + c*HW;
 				for (int p = 0; p < params.P; p++) {
 					for (int q = 0; q < params.Q; q++, out_index++) {
-						for (int h = params.start_indexes_h[p]; h < params.end_indexes_h[p]; h++) {
-							for (int w = params.start_indexes_w[q]; w < params.end_indexes_w[q]; w++) {
+						// Note: We do not treat pad as zero and hence we don't do:  
+						// outputArray[out_index] = Math.max(outputArray[out_index], 0) 
+						// if params.start_indexes_h[p] < 0 || params.end_indexes_h[p] >= params.H || params.start_indexes_w[q] < 0 || params.end_indexes_w[q] >= params.W
```
Comment (dusenberrymw) [inline src/main/java/org/apache/sysml/runtime/matrix/data/LibMatrixDNN.java]: This comment is good to leave for now.

**Pull request closed without merging**